### ISSUE TITLE
Implementing SGD

### ISSUE THREAD

**Issue by smao-astro** [lululxvi/deepxde#175]
Hi Lu,

I am interested in how to implement stochastic gradient descent in DeepXDE for training forward PDE problem.

The word SGD here, refer to:
1. At the beginning of one epoch, generate `N_total` number of points over the domain and on BC. Randomly shuffle them. Divide into `m` batches.
2. For every step,  get a new batch of points, evaluate loss, update weight.
3. Iterate for all the batch.

Currently, as far as I know, DeepXDE does not implement SGD, probably because most of cases gradient descent is enough, but for complex problem, SGD could help.

I read part of the code in the repository, and #39 , and find https://github.com/lululxvi/deepxde/issues/39#issuecomment-626407771 have a discussion, seems that modifying `deepxde.data.pde.PDE.train_next_batch` and `deepxde.model.Model._train_sgd` could achieve the goal, and I am wondering that do you interested in implementing it, or does that on your to-do list.

I also would like to discuss a potential issue here: based on my experience and doc for `deepxde.data.pde.PDE.resample_train_points`: https://github.com/lululxvi/deepxde/blob/e7e5276f63b24ba3e49614fa5cb3e4b47daf6285/deepxde/data/pde.py#L155-L157
I am worried that every update of `deepxde.data.pde.PDE.train_x` would require call of `deepxde.model.Model.compile` which could cause the training very slow, do you think this is true? Could you give some suggestion on how to avoid this? Also I am guessing if this is a issue, it might caused by https://github.com/lululxvi/deepxde/blob/e7e5276f63b24ba3e49614fa5cb3e4b47daf6285/deepxde/data/pde.py#L114
Please confirm me about that and give some suggestion, thanks!

**Comment by lululxvi** [comment]
- Yes, we can follow [#39 (comment)](https://github.com/lululxvi/deepxde/issues/39#issuecomment-626407771) to implement SGD by modifying `deepxde.data.pde.PDE.train_next_batch`. No need to modify `deepxde.model.Model._train_sgd`. Also note that you need to keep the points for each BC and PDE the same in each mini-batch, e.g., always 100 BC points and 1000 PDE points, so you need to shuffle and divide each group of the points separately.
- You can also refer the mini-batch implemented in `OpDataSet`:
https://github.com/lululxvi/deepxde/blob/e7e5276f63b24ba3e49614fa5cb3e4b47daf6285/deepxde/data/op_dataset.py#L54
- One tricky thing here is what you mentioned: the PDE and also the BC may depend on the value of the train points. For example, int the Poisson example, the BC uses Numpy on the training poins:
https://github.com/lululxvi/deepxde/blob/e7e5276f63b24ba3e49614fa5cb3e4b47daf6285/examples/Poisson_Dirichlet_1d.py#L21
This is the reason why we need to rebuild the loss after we update the points. Also, see https://github.com/lululxvi/deepxde/issues/79 for another scenario where it is much more convenient to use Numpy than TensorFlow operations.
- If we never use Numpy and always use TensorFlow on `x`, then there is no need to rebuild. But, because of the previous point, Numpy is much more convenient and flexible in some scenarios. (Maybe later TensorFlow will have all the functions in Numpy...)
- However, it becomes slow if we rebuild each time. To avoid this, a very simple solution is to modify the following line
https://github.com/lululxvi/deepxde/blob/e7e5276f63b24ba3e49614fa5cb3e4b47daf6285/deepxde/data/pde.py#L107
as
```python
tf_random_choice(fi[bcs_start[-1] :], N)
```
Here, we randomly choose `N` items from all the losses using TensorFlow. Similaily for the BC loss. This seems to be the best way. `tf_random_choice` can be found at https://stackoverflow.com/questions/41123879/numpy-random-choice-in-tensorflow, https://github.com/tensorflow/tensorflow/issues/8496#issuecomment-417123890 

But currently I don't have time to implement and (more importantly) test the code.

**Comment by smao-astro** [comment]
Hi Lu,

Thank you for your prompt reply, I have a simple question

> Here, we randomly choose N items from all the losses using TensorFlow. Similaily for the BC loss. This seems to be the best way. tf_random_choice can be found at https://stackoverflow.com/questions/41123879/numpy-random-choice-in-tensorflow, tensorflow/tensorflow#8496 (comment)

For this way, do we need to generate all the training points at some point at the start of the training?
In other word, can we generate the training points on-the-fly, so that we do not need to keep all the training data in the memory and number of samples not limited (I know that we might do not need that much training points in reality, just curious)

**Comment by lululxvi** [comment]
Yes, we need to generate all the training points at the same time and keep them in memory. But storing the points will not consume too much memory. We can regenerate the points every certain epochs, as in the diffusion_1d_resample.py example.

A main issue is that we need to do the forward-propagation for all the points. It needs some modification on this approach.

**Comment by smao-astro** [comment]
Hi Lu,

Thank you for your reply, and I agree with you that maybe we can try to modify https://github.com/lululxvi/deepxde/blob/e7e5276f63b24ba3e49614fa5cb3e4b47daf6285/deepxde/data/pde.py#L107 and I am also currently interested in have a try. However, I find `deepxde.data.pde.PDE.losses` looks a bit messy to me, could you explain a bit more of the logic and the comment of this part of code? Thanks.

**Comment by lululxvi** [comment]
I am glad that you are interested in implementing it. Here are more details about `deepxde.data.pde.PDE.losses`
- When we define pde, the pde can take two arguments `x` and `y`, which is the common case, but pde can also take 3 arguments `x`, `y` and `X` as discussed in #79, or pde can be None (i.e., we don't provide PDE). That is why we have different cases when calling pde
```python
        if self.pde is not None:
            if get_num_args(self.pde) == 2:
                f = self.pde(model.net.inputs, outputs)
            elif get_num_args(self.pde) == 3:
                f = self.pde(model.net.inputs, outputs, self.train_x)
            if not isinstance(f, (list, tuple)):
                f = [f]
```
- The pde may return one residual, or a list of multiple residual (i.e., PDE system). If there is only one PDE, we make it to be a list manually for convenience.
```python
            if not isinstance(f, (list, tuple)):
                f = [f]
```
- We may use the same loss for all the PDE and BCs, or we may use different losses, see an example at #174. If we have only one loss, we again make it to be a list manually.
```python
        if not isinstance(loss, (list, tuple)):
            loss = [loss] * (len(f) + len(self.bcs))
```
- `losses_train()` is the train loss, and `losses_test()` is the test loss. Let us only discuss `losses_train()`.
- The following line just re-compute the PDE if the PDE takes 3 arguments.
```
            f_train = f
            # The following code seems redundant, but if removed, error occurs in losses_test()
            if self.pde is not None and get_num_args(self.pde) == 3:
                f_train = self.pde(model.net.inputs, outputs, self.train_x)
                if not isinstance(f_train, (list, tuple)):
                    f_train = [f_train]
```
- See the comment here https://github.com/lululxvi/deepxde/issues/39#issuecomment-626407771 and also the docstring of `deepxde.data.pde.PDE` for the order of `train_x`. See the docstring of `deepxde.data.pde.PDE` for `self.num_bcs` . We compute the starting index of the training points for each BC:
```python
            bcs_start = np.cumsum([0] + self.num_bcs)
```
- `bcs_start[-1]` is the staring index for the pde loss. We compute the pde (system) residuals `error_f`, and them compute the loss by comparing with 0. Note: as discussed before, `loss` is a list, and `loss[i]` is the loss function for i-th residual.
```python
            error_f = [fi[bcs_start[-1] :] for fi in f_train]
            losses = [
                loss[i](tf.zeros(tf.shape(error), dtype=config.real(tf)), error)
                for i, error in enumerate(error_f)
            ]
```
- We compute the BC loss. `bc.error` computes the error for each bc.
```python
            for i, bc in enumerate(self.bcs):
                beg, end = bcs_start[i], bcs_start[i + 1]
                error = bc.error(self.train_x, model.net.inputs, outputs, beg, end)
                losses.append(
                    loss[len(error_f) + i](
                        tf.zeros(tf.shape(error), dtype=config.real(tf)), error
                    )
                )
```

Let me know if this is clear.

**Comment by smao-astro** [comment]
Hi Lu,

I went to explore a efficient way of resampling every step without call `model.compile`, in order to implement so, I tried to replace every occurrence of numpy object in tensor graph by tensor and tf ops, for example, I managed to replace `deepxde.geometry.geometry_1d.Interval.boundary_normal` and `deepxde.boundary_conditions.DirichletBC.error` from `def error(self, X, inputs, outputs, beg, end)` to `def error(self, inputs, outputs, beg, end)`, thus the code won't DIRECTLY use any numpy array when build the graph.

However the code still indirectly using the numpy array building the graph: `deepxde.data.pde.PDE` is using `num_bcs` to divide inputs, outputs and thus losses into different parts, and because of the mechanism of static graph mode, `num_bcs` won't be updated during the run (but points does, especially when randomly sampling inside domain and on boundary), so when the number of inputs for bcs changed, points for one bc will be feed to another bc.

I currently do not have a good way to resolve this, thus find it a bit tricky to implement a robust and efficient mini-batch gradient training.

Do you have any idea? Thanks.

**Comment by lululxvi** [comment]
I think we can just keep that each time the number of points for each BC are the same, i.e., `num_bcs` doesn't change. Also, actually I think we can just use whole batch for BC/IC, and only resample the points inside the domain, because the points inside are much more (2 orders of magnitude more) than BC points.

**Comment by smao-astro** [comment]
I agree with you, actually I have modified `random_boundary_points` to try to make sure the number of points on bc is fixed, however the current implementation of random sample such as `deepxde.geometry.geometry_nd.Hypercube.random_points` would add additional points to boundary, especially when `n` is large, which is the actual case of changes of `num_bcs` over steps.

**Comment by lululxvi** [comment]
The `random_points` function is supported to only sample random points inside the domain, not on the boundary. In which case, do you find that it would sample points on the boundary? I should fix this.

**Comment by smao-astro** [comment]
The current implementation
https://github.com/lululxvi/deepxde/blob/c0da2a4b1db25f2e383dcd63334dda6a3920ea38/deepxde/geometry/geometry_nd.py#L68-L74
DOES  allowed to sample points very close to boundary, though the probability is small. Try
```python
n = 10 ** 6
xmin = 0.4
xmax = 2.5
x = np.random.default_rng().random(n)
x = (xmax - xmin) * x + xmin
print(np.sum(
    np.isclose(x, xmin)
))
```

**Comment by lululxvi** [comment]
Ahh, this is a numerical issue. You are right: we need to do something for this.

**Comment by smao-astro** [comment]
Do you have any idea? If no, would you like to go to check whether it is globally possible merge `pde.bc_points()` and `pde.train_points()` in `deepxde.data.pde.PDE.train_next_batch` so that https://github.com/lululxvi/deepxde/blob/9d460afafe0cd723cc1c767f233756bef5480628/deepxde/data/pde.py#L212 only apply on points from https://github.com/lululxvi/deepxde/blob/9d460afafe0cd723cc1c767f233756bef5480628/deepxde/data/pde.py#L193-L199

Thanks.

**Comment by lululxvi** [comment]
In addition to `self.train_x_all`, we can have another variable `self.train_x_bc` for all BC points. `self.train_x_bc` include all the points for `self.num_boundary` and the BC points in `self.anchors`.

**Comment by smao-astro** [comment]
I see.
